src/repr/src/proto/scalar.proto

```diff
@@ -0,0 +1,84 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+syntax = "proto3";
+
+package scalar;
+
+import "google/protobuf/empty.proto";
+import "adt/char.proto";
+import "adt/numeric.proto";
+import "adt/varchar.proto";
+
+message ProtoScalarType {
+    // ScalarType::Numeric contains an Option<NumericMaxScale>
+    // None is encoded as this field not present.
+    message ProtoNumeric {
+        adt.numeric.ProtoNumericMaxScale max_scale = 1;
+    }
+    message ProtoChar {
+        adt.char.ProtoCharLength length = 1;
+    }
+
+    message ProtoVarChar {
+        adt.varchar.ProtoVarCharMaxLength max_length = 1;
+    }
+
+    message ProtoList {
+        ProtoScalarType element_type = 1;
+        optional uint32 custom_oid = 2;
+    }
+
+    message ProtoRecordField {
+        string ColumnName = 1; // TODO: Create and use ProtoColumnName
+        string ColumnType = 2; // TODO: Create and use ProtoColumnType
+    }
+
+    message ProtoRecord {
+        repeated ProtoRecordField fields = 1;
+        optional uint32 custom_oid = 2;
+        optional string custom_name = 3;
+    }
+
+    message ProtoMap {
+        ProtoScalarType value_type = 1;
+        optional uint32 custom_oid = 2;
+    }
+
+    oneof kind  {
```

Comment:
You might want to use the number 1-15 for the most frequent variants: https://developers.google.com/protocol-buffers/docs/proto3#assigning_field_numbers

Reply:
Or, only use half to have some future flexibility.